File: wasm-core/tests/compute.rs
use pulseview_wasm_core::{
    compute_frequency_points, compute_histogram_counts, compute_histogram_meta,
    compute_stats_values, count_pulses_from_transitions, derive_falling_edges, derive_rising_edges,
    invert_transition_levels,
};

#[test]
fn derives_edges_and_counts_complete_high_pulses() {
    let times = [0.0, 1.0, 2.0, 4.0, 5.0, 6.0];
    let levels = [0, 1, 0, 1, 0, 1];
    assert_eq!(derive_rising_edges(&times, &levels), vec![1.0, 4.0, 6.0]);
    assert_eq!(derive_falling_edges(&times, &levels), vec![2.0, 5.0]);
    assert_eq!(count_pulses_from_transitions(&levels), 2);
    let inverted = invert_transition_levels(&levels);
    assert_eq!(inverted, vec![1, 0, 1, 0, 1, 0]);
    assert_eq!(derive_falling_edges(&times, &inverted), vec![1.0, 4.0, 6.0]);
}

#[test]
fn computes_sample_statistics() {
    assert!(compute_stats_values(&[]).is_empty());
    assert_eq!(compute_stats_values(&[4.0]), vec![4.0, 4.0, 4.0, 0.0, 0.0]);
    let stats = compute_stats_values(&[1.0, 2.0, 3.0]);
    assert_eq!(&stats[..3], &[1.0, 3.0, 2.0]);
    assert!((stats[3] - 1.0).abs() < 1e-12);
    assert!((stats[4] - 50.0).abs() < 1e-12);
}

#[test]
fn computes_histogram_metadata_and_counts() {
    assert!(compute_histogram_meta(&[], 10, 80).is_empty());
    assert!(compute_histogram_meta(&[3.0, 3.0], 10, 80).is_empty());
    let values = [0.0, 1.0, 2.0, 3.0];
    let meta = compute_histogram_meta(&values, 2, 4);
    assert_eq!(meta, vec![0.0, 3.0, 4.0]);
    assert_eq!(
        compute_histogram_counts(&values, 0.0, 3.0, 4),
        vec![1, 1, 1, 1]
    );
}

#[test]
fn histogram_matches_typescript_non_finite_and_clamp_semantics() {
    // NaN 频率在 TypeScript 中不会累加任何数值桶。
    let with_nan = [1.0, f64::NAN, 2.0, 3.0];
    assert_eq!(
        compute_histogram_counts(&with_nan, 1.0, 3.0, 4),
        vec![1, 0, 1, 1]
    );
    // +Infinity 在 TypeScript 中落入最后一桶。
    let with_infinity = [1.0, f64::INFINITY, 2.0, 3.0];
    assert_eq!(
        compute_histogram_counts(&with_infinity, 1.0, 3.0, 4),
        vec![1, 0, 1, 2]
    );
    // 负数 minBins/maxBins 与 Math.max(min, Math.min(max, ceil)) 语义一致：
    // 默认最小 10 桶时，minBins=10、maxBins=-5 仍取 10。
    let values = [0.0, 1.0, 2.0, 3.0];
    let meta = compute_histogram_meta(&values, 10, -5);
    assert_eq!(meta, vec![0.0, 3.0, 10.0]);
    // minBins 为负数时退化为 ceil 结果：ceil(1 + 3.322*log10(4)) = 4。
    let meta_negative_min = compute_histogram_meta(&values, -5, 80);
    assert_eq!(meta_negative_min, vec![0.0, 3.0, 4.0]);
    // 非正 bin_count 不分配内存，直接返回空结果。
    assert!(compute_histogram_counts(&values, 0.0, 3.0, 0).is_empty());
    assert!(compute_histogram_counts(&values, 0.0, 3.0, -2).is_empty());
}

#[test]
fn pulse_mode_emits_nan_period_when_fall_edge_missing_like_typescript() {
    // levels = [0,1,1,0,1,1]：rises=[1,4]，falls=[3]。
    // 第 2 个脉冲（k=1）的 falls[k] 越界，TypeScript 得到 NaN 周期；
    // 不能回退成有限值。
    let times = [0.0, 1.0, 2.0, 3.0, 4.0, 5.0];
    let levels = [0, 1, 1, 0, 1, 1];
    let points = compute_frequency_points(&times, &levels, 0, true, 0, false, 0.0, 0);
    assert_eq!(points.len(), 8); // 2 个点 × 4 字段
                                 // 第一个点：time=t[1]=1，freq=1/period=1/2=0.5，period 字段=脉宽 w=1，duty=0.5
    assert_eq!(points[0], 1.0);
    assert_eq!(points[1], 0.5);
    assert_eq!(points[2], 1.0);
    assert_eq!(points[3], 0.5);
    // 第二个点：time=t[4]=4，freq=1/NaN=NaN，period 字段=脉宽 w=1，duty=0.5
    assert_eq!(points[4], 4.0);
    assert!(points[5].is_nan(), "freq must be NaN, got {}", points[5]);
    assert_eq!(points[6], 1.0);
    assert_eq!(points[7], 0.5);
}

#[test]
fn low_gap_mode_emits_nan_point_when_rise_exhausted_like_typescript() {
    // levels = [0,1,0,0,0,2,1,1,0,0,2,0]：rises=[1]，falls=[2,8]。
    // k=1 时 rises[k-f0] 越界，TypeScript 产出 freq=NaN、period=6、duty=NaN。
    let times: Vec<f64> = (-500..-488).map(|i| i as f64).collect();
    let levels = [0, 1, 0, 0, 0, 2, 1, 1, 0, 0, 2, 0];
    let points = compute_frequency_points(&times, &levels, 3, false, 0, false, 0.0, 0);
    assert_eq!(points.len(), 4); // 1 个点
    assert_eq!(points[0], -495.0);
    assert!(points[1].is_nan(), "freq must be NaN");
    assert_eq!(points[2], 6.0);
    assert!(points[3].is_nan(), "duty must be NaN");
}

File: src/wasm/compute.ts
import type { FreqPoint } from '../types.ts';
import type { LowGapMarker } from '../types.ts';
import { assertIntegerArraysEquivalent, assertNumericArraysEquivalent, dualRun } from './compare.ts';
import { getWasmExports } from './runtime.ts';

export interface ComputeReferences {
  deriveEdgesFromTransitions(times: Float64Array, levels: Int8Array): {
    risingEdges: Float64Array;
    fallingEdges: Float64Array;
  };
  invertTransitionLevels(levels: Int8Array): Int8Array;
  countPulsesFromTransitions(levels: Int8Array): number;
  computeStats(points: FreqPoint[]): {
    min: number;
    max: number;
    avg: number;
    std: number;
    cv: number;
  } | null;
}

const frequencyModeIds = { pulse: 0, rising: 1, falling: 2, 'low-gap': 3 } as const;

function unflattenPoints(values: Float64Array): FreqPoint[] {
  const result: FreqPoint[] = [];
  for (let index = 0; index < values.length; index += 4) {
    result.push({
      time: values[index],
      freq: values[index + 1],
      period: values[index + 2],
      dutyCycle: Number.isNaN(values[index + 3]) ? undefined : values[index + 3],
    });
  }
  return result;
}

function assertPointsEquivalent(moduleName: string, inputSize: number, ts: FreqPoint[], wasm: FreqPoint[]): void {
  if (ts.length !== wasm.length) throw new Error(`length TS=${ts.length}, WASM=${wasm.length}`);
  for (let index = 0; index < ts.length; index++) {
    const left = ts[index];
    const right = wasm[index];
    assertNumericArraysEquivalent(moduleName, inputSize,
      [left.time, left.freq, left.period ?? Number.NaN, left.dutyCycle ?? Number.NaN],
      [right.time, right.freq, right.period ?? Number.NaN, right.dutyCycle ?? Number.NaN]);
  }
}

export function wasmDeriveEdges(
  times: Float64Array,
  levels: Int8Array,
  references: ComputeReferences,
) {
  const wasm = getWasmExports();
  if (!wasm) return references.deriveEdgesFromTransitions(times, levels);
  return dualRun(
    'deriveEdgesFromTransitions',
    times.length,
    () => references.deriveEdgesFromTransitions(times, levels),
    () => ({
      risingEdges: wasm.derive_rising_edges(times, levels),
      fallingEdges: wasm.derive_falling_edges(times, levels),
    }),
    (ts, result) => {
      assertNumericArraysEquivalent('risingEdges', times.length, ts.risingEdges, result.risingEdges);
      assertNumericArraysEquivalent('fallingEdges', times.length, ts.fallingEdges, result.fallingEdges);
    },
  );
}

export function wasmInvertLevels(levels: Int8Array, references: ComputeReferences): Int8Array {
  const wasm = getWasmExports();
  if (!wasm) return references.invertTransitionLevels(levels);
  return dualRun(
    'invertTransitionLevels', levels.length,
    () => references.invertTransitionLevels(levels),
    () => wasm.invert_transition_levels(levels),
    (ts, result) => assertIntegerArraysEquivalent('invertTransitionLevels', levels.length, ts, result),
  );
}

export function wasmCountPulses(levels: Int8Array, references: ComputeReferences): number {
  const wasm = getWasmExports();
  if (!wasm) return references.countPulsesFromTransitions(levels);
  return dualRun(
    'countPulsesFromTransitions', levels.length,
    () => references.countPulsesFromTransitions(levels),
    () => wasm.count_pulses_from_transitions(levels),
    (ts, result) => {
      if (ts !== result) throw new Error(`TS=${ts}, WASM=${result}`);
    },
  );
}

export function wasmComputeStats(points: FreqPoint[], references: ComputeReferences) {
  const wasm = getWasmExports();
  if (!wasm) return references.computeStats(points);
  const frequencies = Float64Array.from(points, (point) => point.freq);
  return dualRun(
    'computeStats', points.length,
    () => references.computeStats(points),
    () => {
      const values = wasm.compute_stats_values(frequencies);
      return values.length === 0
        ? null
        : { min: values[0], max: values[1], avg: values[2], std: values[3], cv: values[4] };
    },
    (ts, result) => {
      if (ts === null || result === null) {
        if (ts !== result) throw new Error(`TS=${ts}, WASM=${result}`);
        return;
      }
      assertNumericArraysEquivalent(
        'computeStats', points.length,
        [ts.min, ts.max, ts.avg, ts.std, ts.cv],
        [result.min, result.max, result.avg, result.std, result.cv],
      );
    },
  );
}

export function wasmComputeHistogram(
  frequencies: number[],
  minBins: number,
  maxBins: number,
  tsRun: () => { labels: string[]; bins: number[] } | null,
) {
  const wasm = getWasmExports();
  if (!wasm) return tsRun();
  const values = Float64Array.from(frequencies);
  return dualRun(
    'computeHistogramBins', frequencies.length, tsRun,
    () => {
      const meta = wasm.compute_histogram_meta(values, minBins, maxBins);
      if (meta.length === 0) return null;
      const [minimum, maximum, rawCount] = meta;
      const binCount = Math.trunc(rawCount);
      const width = (maximum - minimum) / binCount;
      // 与 TypeScript 的 new Array(binCount) 一致：负数桶数同样抛 RangeError，
      // 而不是被 Array.from 静默截断为 0。
      const labels = new Array<string>(binCount);
      for (let index = 0; index < binCount; index++) {
        const frequency = minimum + (index + 0.5) * width;
        if (frequency >= 1e9) labels[index] = `${(frequency / 1e9).toFixed(2)}G`;
        else if (frequency >= 1e6) labels[index] = `${(frequency / 1e6).toFixed(2)}M`;
        else if (frequency >= 1e3) labels[index] = `${(frequency / 1e3).toFixed(2)}k`;
        else labels[index] = frequency.toFixed(2);
      }
      return { labels, bins: Array.from(wasm.compute_histogram_counts(values, minimum, maximum, binCount)) };
    },
    (ts, result) => {
      if (ts === null || result === null) {
        if (ts !== result) throw new Error(`TS=${ts}, WASM=${result}`);
        return;
      }
      if (ts.labels.join('\0') !== result.labels.join('\0')) throw new Error('histogram labels differ');
      assertIntegerArraysEquivalent('computeHistogramBins', frequencies.length, ts.bins, result.bins);
    },
  );
}

export function wasmComputeFrequency(
  times: Float64Array,
  levels: Int8Array,
  mode: keyof typeof frequencyModeIds,
  dutyCorrect: boolean,
  edgeBase: 'falling' | 'rising',
  toleranceEnabled: boolean,
  tolerancePct: number,
  defaultLevel: 0 | 1,
  tsRun: () => FreqPoint[],
): FreqPoint[] {
  const wasm = getWasmExports();
  if (!wasm) return tsRun();
  return dualRun(
    'computeFreqFromTransitions', times.length, tsRun,
    () => unflattenPoints(wasm.compute_frequency_points(
      times, levels, frequencyModeIds[mode], dutyCorrect,
      edgeBase === 'rising' ? 1 : 0, toleranceEnabled, tolerancePct, defaultLevel,
    )),
    (ts, result) => assertPointsEquivalent('computeFreqFromTransitions', times.length, ts, result),
  );
}

export function wasmComputeLowGapMarkers(
  times: Float64Array,
  levels: Int8Array,
  threshold: number,
  toleranceEnabled: boolean,
  tolerancePct: number,
  tsRun: () => LowGapMarker[],
): LowGapMarker[] {
  const wasm = getWasmExports();
  if (!wasm) return tsRun();
  return dualRun(
    'computeLowGapMarkers', times.length, tsRun,
    () => {
      const flat = wasm.compute_low_gap_markers(times, levels, threshold, toleranceEnabled, tolerancePct);
      const result: LowGapMarker[] = [];
      for (let index = 0; index < flat.length; index += 4) {
        result.push({ startTime: flat[index], endTime: flat[index + 1], gap: flat[index + 2], dutyCycle: flat[index + 3] });
      }
      return result;
    },
    (ts, result) => {
      if (ts.length !== result.length) throw new Error(`length TS=${ts.length}, WASM=${result.length}`);
      for (let index = 0; index < ts.length; index++) {
        assertNumericArraysEquivalent('computeLowGapMarkers', times.length,
          [ts[index].startTime, ts[index].endTime, ts[index].gap, ts[index].dutyCycle],
          [result[index].startTime, result[index].endTime, result[index].gap, result[index].dutyCycle]);
      }
    },
  );
}
